src/auth.rs

```diff
@@ -198,14 +198,22 @@ where
     // All the preamble is done, now receive username+password.
     let (username, password) = client_username_password(ws_stream).await?;
 
-    match validate_token(&username, &replid, pubkeys).context("token validation") {
-        Ok(()) => {}
-        Err(err) => {
-            // Let the client know that it messed up.
-            stream.write_all(b"\x00\x00\x00\x01").await?;
-            return Err(err);
-        }
-    };
+    // We can let users through if they use the "runner" username, _but_ that's only if a password
+    // is provided _and_ the server requires a password.
+    let mut used_token = false;
+    if username != "runner" || password.is_empty() {
+        match validate_token(&username, &replid, pubkeys).context("token validation") {
```

Comment:
This might be slightly more idiomatic:

```rust
if let Err(err) = validate_token(&username, &replid, pubkeys).context("token validation") {
      // Let the client know that it messed up.
      ws_stream
          .send(WebSocketMessage::Binary((b"\x00\x00\x00\x01").to_vec()))
          .await?;
      return Err(err);
};
```